Correct implementation of BinaryName and InternalForm

gsaga opened the pull request:
This pull request addresses the issue #2098 and adds rules for type deduction of `String.replace` return types.

mernst commented:
Could you please fix the build failures?
One shows crashes in the checker.
The other is stylistic.

mernst commented:
Please also pull master into your branch.  You don't seem to have done that for more than a month.

gsaga commented:
This is the error build 8244.3 is emitting:

error: MultiGraphQualifierHierarchy could not determine GLB for
@org.checkerframework.checker.signature.qual.BinaryNameForNonArrayInUnnamedPackage
and @org.checkerframework.checker.signature.qual.ClassGetSimpleName. Please
ensure that the checker knows about all type qualifiers.

It's weird because I removed the annotation
'BinaryNameForNonArrayInUnnamed' package in the pull request as you can
sees:
https://github.com/gsaga/checker-framework/tree/fix_binary_name_internal_form/checker/src/main/java/org/checkerframework/checker/signature/qual


On Tue, Aug 7, 2018 at 11:46 PM Michael Ernst <notifications@github.com>
wrote:

> Could you please fix the build failures?
> One shows crashes in the checker.
> The other is stylistic.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/typetools/checker-framework/pull/2100#issuecomment-411152016>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AU-teUXuQOh-dsWLvmtQ745KFIXYafJBks5uOdl1gaJpZM4VylEE>
> .
>


-- 
Best Regards
Sagar Tewari

gsaga commented:
I've created a pull request in plumelib repository which replaces BinaryNameForNonArray with BinaryName
https://github.com/mernst/plume-lib/pull/56

mernst commented:
Job 6 in the CI build is failing because of missing documentation.  Please pull master again to resolve it.

mernst commented on checker/src/main/java/org/checkerframework/checker/signature/qual/SourceNameForNonArrayNonInner.java:
```diff
@@ -14,6 +15,7 @@
+@SubtypeOf({FullyQualifiedName.class, BinaryName.class})
+@ImplicitFor(stringPatterns = "^[A-Za-z_][A-Za-z_0-9]*(\\.[A-Za-z_][A-Za-z_0-9]*)*$")
 @Target({ElementType.TYPE_USE, ElementType.TYPE_PARAMETER})
 public @interface SourceNameForNonArrayNonInner {}
```
This is an ugly name, and one that isn't intuitive to users.  (That isn't your fault!  The name predates your work.)
Let's rename this to `DotSeparatedIdentifiers`.  Don't forget to update the Javadoc too.  Thanks.

mernst commented on checker/src/main/java/org/checkerframework/checker/signature/SignatureAnnotatedTypeFactory.java:
```diff
@@ -29,22 +30,32 @@
     private final ExecutableElement replaceCharSequenceCharSequence;
 
+    /** The {@link String#contains(CharSequence)} method. */
+    private final ExecutableElement containsCharSequence;
```
Please name this `stringContains`, because the containing class is more important than the parameter type and the `contains` method is not overloaded.
`replaceCharChar` contains its parameter types because the `replace` method is overloaded.

mernst commented on checker/src/main/java/org/checkerframework/checker/signature/SignatureAnnotatedTypeFactory.java:
```diff
@@ -106,47 +124,65 @@ public Void visitCompoundAssignment(CompoundAssignmentTree node, AnnotatedTypeMi
         }
+
+        /**
+         * Handles String.replace calls when the receiver is a {@link
```
Please give a more specific description than "handles", which doesn't say what this does.

Once we have a nice library for doing conversions among string formats, we could get rid of all this code and direct users to use the library instead, which will be clearer and less error-prone.

mernst commented on checker/tests/signature/Conversion.java:
```diff
@@ -40,6 +40,14 @@
+            return bn.replace('$', '.');
+        }
+
+        @FullyQualifiedName String fullyQualifiedNameToDotSeparatedIdentifiers(@BinaryName String bn) {
```
The method name is not consistent with the formal parameter types.

mernst commented on checker/tests/signature/Conversion.java:
```diff
@@ -80,5 +88,13 @@
+            return bn.replace("$", ".");
+        }
+
+        @FullyQualifiedName String fullyQualifiedNameToDotSeparatedIdentifiers(@BinaryName String bn) {
```
The method name is not consistent with the formal parameter types.

mernst commented on checker/tests/signature/SignatureTypeFactoryTest.java:
```diff
@@ -215,7 +215,7 @@ void m() {
         // :: error: (assignment.type.incompatible)
         fqn = s12;
-        bn = s12;
+        // bn = s12; // s12 isn't a valid BinaryName
```
Please don't comment out tests.  If an error is expected, write a `// ::` comment on the preceding line.

mernst commented on checker/src/main/java/org/checkerframework/checker/signature/qual/FieldDescriptorForPrimitive.java:
```diff
@@ -0,0 +1,20 @@
+ * href="https://docs.oracle.com/javase/specs/jvms/se8/html/jvms-4.html#jvms-4.3.2">Java Virtual
+ * Machine Specification, section 4.3.2</a>.
+ *
+ * <p>Must be one of B,C,D,F,I,J,S,Z
```
Please end each sentence with a period, and put a space after each comma.

mernst commented on checker/src/main/java/org/checkerframework/checker/signature/qual/FieldDescriptorForPrimitive.java:
```diff
@@ -0,0 +1,20 @@
+ *
+ * @checker_framework.manual #signature-checker Signature Checker
+ */
+@SubtypeOf({FieldDescriptorForArray.class, Identifier.class})
```
Seeing `FieldDescriptorForArray` is surprising -- a primitive is not an array.  I think the regex for `FieldDescriptorForArray` should not match primitives, and the name `FieldDescriptorForArray` should be changed to `FieldDescriptorForArrayWithoutPackage`.  If there is a need for the concept currently represented by `FieldDescriptorForArray` (which includes non-arrays), then it should be given a different name.

mernst commented:
This is looking good and is nearly ready to merge.

In file `checker/tests/signature/SignatureTypeFactoryTest.java`, the text at
```
    // The hierarchy of type representations contains:
```
is out of date.  Please search the entire repositor for mentions of the types that you removed, and update them.

The diagram in figure 15.1 (file docs/manual/signature-types.svg).
It's a bit of a pain to update, but now is the time since the type hierarchy is finalized.

mernst commented:
I would like to keep all the type hierarchy diagrams in the manual looking conistent.
(It's great that your diagram is automatically generated, but the text in it is too small to read; manual layout is better for the manual.)
File `docs/manual/figures/README` says how to edit the `.svg` directly; I use Inkscape.

gsaga commented:
I wasn't talking about the pdf file I attached(that's quite ugly). The file
attached to this mail is a https://www.draw.io diagram exported to svg.
It's specifically designed for making diagrams similar to the ones in
checkerframework manual. I've made a query at the forum if this tool
provides an api to programmatically export files.


On Sun, Aug 12, 2018 at 10:21 PM Michael Ernst <notifications@github.com>
wrote:

> I would like to keep all the type hierarchy diagrams in the manual looking
> conistent.
> (It's great that your diagram is automatically generated, but the text in
> it is too small to read; manual layout is better for the manual.)
> File docs/manual/figures/README says how to edit the .svg directly; I use
> Inkscape.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/typetools/checker-framework/pull/2100#issuecomment-412355897>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AU-tef3SYrKhmbOM1bHgkyI9dtgQ7RZoks5uQF0WgaJpZM4VylEE>
> .
>


-- 
Best Regards
Sagar Tewari

gsaga commented:
This is how the locally built documentation looks like.

On Sun, Aug 12, 2018 at 10:34 PM Sagar Tewari <iaansagar@gmail.com> wrote:

> I wasn't talking about the pdf file I attached(that's quite ugly). The
> file attached to this mail is a https://www.draw.io diagram exported to
> svg. It's specifically designed for making diagrams similar to the ones in
> checkerframework manual. I've made a query at the forum if this tool
> provides an api to programmatically export files.
>
>
> On Sun, Aug 12, 2018 at 10:21 PM Michael Ernst <notifications@github.com>
> wrote:
>
>> I would like to keep all the type hierarchy diagrams in the manual
>> looking conistent.
>> (It's great that your diagram is automatically generated, but the text in
>> it is too small to read; manual layout is better for the manual.)
>> File docs/manual/figures/README says how to edit the .svg directly; I
>> use Inkscape.
>>
>> —
>> You are receiving this because you authored the thread.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/typetools/checker-framework/pull/2100#issuecomment-412355897>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AU-tef3SYrKhmbOM1bHgkyI9dtgQ7RZoks5uQF0WgaJpZM4VylEE>
>> .
>>
>
>
> --
> Best Regards
> Sagar Tewari
>


-- 
Best Regards
Sagar Tewari

mernst commented:
We're having this discussion at https://github.com/typetools/checker-framework/pull/2100, not by email.  I don't see an attachment.  Can you please attach the file to the GitHub issue or email it to me directly?  Thanks.

mernst commented:
PS: In general, it's best not to reply to GitHub issue conversations by email, even though that is convenient and tempting. :-)

gsaga commented:
I've updated the README with instructions.

On Mon, Aug 13, 2018 at 8:53 PM Michael Ernst <notifications@github.com>
wrote:

> *@mernst* approved this pull request.
>
> This looks good. Thanks!
>
> One more thing, then we can merge.
> Please add text to file manual/figures/README indicating exactly how you
> created file lignature-types.svg. It's OK if some of these are manual
> steps, but make sure that others can update the figure in the future.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/typetools/checker-framework/pull/2100#pullrequestreview-145705541>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AU-tedlp5xzz-G7Fxq_0nRHAcDpWPgwAks5uQZnpgaJpZM4VylEE>
> .
>


-- 
Best Regards
Sagar Tewari

mernst commented:
The tests are failing:
https://travis-ci.org/typetools/checker-framework/jobs/415502623#L1349
Can you please fix that?

Also, once this is merged, the plume-lib jobs will start to fail.
Can you please add links to the relevant branches, which pass your new type-checker, here in the conversation?  That will enable us to merge them quickly after merging this pull request, reducing the total breakage time.

With those two changes, this is ready to merge.

gsaga commented:
I've updated the pull request with the fix for that.

On Tue, Aug 14, 2018 at 1:01 AM Michael Ernst <notifications@github.com>
wrote:

> The tests are failing:
> https://travis-ci.org/typetools/checker-framework/jobs/415502623#L1349
> Can you please fix that?
>
> Also, once this is merged, the plume-lib jobs will start to fail.
> Can you please add links to the relevant branches, which pass your new
> type-checker, here in the conversation? That will enable us to merge them
> quickly after merging this pull request, reducing the total breakage time.
>
> With those two changes, this is ready to merge.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/typetools/checker-framework/pull/2100#issuecomment-412636430>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AU-teWbjPYV0koM0m54CtDq2pjn7qk9Qks5uQdQUgaJpZM4VylEE>
> .
>


-- 
Best Regards
Sagar Tewari

mernst commented on checker/src/main/java/org/checkerframework/checker/signature/SignatureAnnotatedTypeFactory.java:
```diff
@@ -157,6 +157,9 @@ public Void visitMethodInvocation(MethodInvocationTree tree, AnnotatedTypeMirror
                     type.replaceAnnotation(BINARY_NAME);
+                } else if ((oldChar == '$' && newChar == '.')
+                        && receiverType.getAnnotation(BinaryName.class) != null) {
+                    type.replaceAnnotation(DOT_SEPARATED_IDENTIFIERS);
```
This is incorrect.  "MyClass$22" is a binary name, but "MyClass.22" is not a dot-separated identifier.
Please remove this conversion rule.

(Actually, this is a bug that I discovered in code review and fixed in commit 084e251629.   For efficiency, I don't point out every issue in the code to you.)

gsaga commented:
I've added 4 error qualifiers to the signature checker tests which check
conversion between BinaryName and DotSeparatedIdentifiers.

On Tue, Aug 14, 2018 at 5:58 AM Michael Ernst <notifications@github.com>
wrote:

> *@mernst* commented on this pull request.
> ------------------------------
>
> In
> checker/src/main/java/org/checkerframework/checker/signature/SignatureAnnotatedTypeFactory.java
> <https://github.com/typetools/checker-framework/pull/2100#discussion_r209799135>
> :
>
> > @@ -157,6 +157,9 @@ public Void visitMethodInvocation(MethodInvocationTree tree, AnnotatedTypeMirror
>                  } else if ((oldChar == '/' && newChar == '.')
>                          && receiverType.getAnnotation(InternalForm.class) != null) {
>                      type.replaceAnnotation(BINARY_NAME);
> +                } else if ((oldChar == '$' && newChar == '.')
> +                        && receiverType.getAnnotation(BinaryName.class) != null) {
> +                    type.replaceAnnotation(DOT_SEPARATED_IDENTIFIERS);
>
> This is incorrect. "MyClass$22" is a binary name, but "MyClass.22" is not
> a dot-separated identifier.
> Please remove this conversion rule.
>
> (Actually, this is a bug that I discovered in code review and fixed in
> commit 084e251
> <https://github.com/typetools/checker-framework/commit/084e2516292aab9d9eb8ee2c8a9d3b17891c2e98>.
> For efficiency, I don't point out every issue in the code to you.)
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/typetools/checker-framework/pull/2100#pullrequestreview-145882970>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AU-tecNdnFcB4prrdd89QLDEGd-aP8toks5uQhm-gaJpZM4VylEE>
> .
>


-- 
Best Regards
Sagar Tewari

gsaga commented:
Hi,

I've made pull requests to plume-util(
https://github.com/plume-lib/plume-util/pull/2) and options (
https://github.com/plume-lib/options/pull/4). bcel-util also needs to be
modified, but since it depends on annotated fork of
bcel(typetools/commons-bcel), it will have to be modified first. I'm
currently working on changing the annotations in commons-bcel.

On Tue, Aug 14, 2018 at 6:54 AM Michael Ernst <notifications@github.com>
wrote:

> Great, thanks. Now it just remains to get the plume-lib libraries
> type-checking without errors (please make links to pull requests), and
> we'll merge this pull request!
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/typetools/checker-framework/pull/2100#issuecomment-412721526>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AU-tefJlLcxTBFhq7CIWX3MqVZZKVq-Zks5uQibKgaJpZM4VylEE>
> .
>


-- 
Best Regards
Sagar Tewari

mernst commented:
Perfect.  Thanks for doing that.  The two pull requests you linked to look good, thanks.
I think you may also need to adapt https://github.com/plume-lib/signature-util ; that is the last of the 4 packages in plume-util that use the Signature Checker.

gsaga commented:
I've made a pull request to signature-util package as well:
https://github.com/plume-lib/signature-util/pull/1

On Wed, Aug 15, 2018 at 1:27 AM Michael Ernst <notifications@github.com>
wrote:

> Perfect. Thanks for doing that. The two pull requests you linked to look
> good, thanks.
> I think you may also need to adapt
> https://github.com/plume-lib/signature-util ; that is the last of the 4
> packages in plume-util that use the Signature Checker.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/typetools/checker-framework/pull/2100#issuecomment-412995120>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AU-teeI8OK9ijb6w79omFjTeBZtbH1Zgks5uQyuogaJpZM4VylEE>
> .
>


-- 
Best Regards
Sagar Tewari

gsaga commented:
These are the pull requests made on user libraries of signature checker which make them compatible with this pull request:

1. https://github.com/plume-lib/plume-util/pull/2

2. https://github.com/plume-lib/options/pull/4

3. https://github.com/plume-lib/signature-util/pull/1

4. https://github.com/typetools/commons-bcel/pull/8

5. https://github.com/gsaga/bcel-util
